### DISCUSSION TITLE
[Multiple USB CDC] - Configure 5 usb cdc devices

### DISCUSSION THREAD

**Discussion by nhhiep92** [hathach/tinyusb#961] [Q&A] [answered]
Hi all,

I'm trying to setup 5 usb cdc devices using rp2040. My pi 3 detected and list down all the device but it only can write/read to 4 port. The last port can not write/read. Please help me to review my setting base on cdc_dual_ports examples. Below is my settings.

#define PICO_EPNUM_CDC_0_NOTIF (0x81)
#define PICO_EPNUM_CDC_0_DATA (0x02)
#define PICO_EPNUM_CDC_1_NOTIF (0x83)
#define PICO_EPNUM_CDC_1_DATA (0x04)
#define PICO_EPNUM_CDC_2_NOTIF (0x85)
#define PICO_EPNUM_CDC_2_DATA (0x06)
#define PICO_EPNUM_CDC_3_NOTIF (0x87)
#define PICO_EPNUM_CDC_3_DATA (0x08)
#define PICO_EPNUM_CDC_4_NOTIF (0x89)
#define PICO_EPNUM_CDC_4_DATA (0x0A)

enum 
{
    PICO_USBD_ITF_CDC_0 = 0,
    PICO_USBD_ITF_CDC_0_DATA,
    PICO_USBD_ITF_CDC_1,
    PICO_USBD_ITF_CDC_1_DATA,
    PICO_USBD_ITF_CDC_2,
    PICO_USBD_ITF_CDC_2_DATA,
    PICO_USBD_ITF_CDC_3,
    PICO_USBD_ITF_CDC_3_DATA,
    PICO_USBD_ITF_CDC_4,
    PICO_USBD_ITF_CDC_4_DATA,
    PICO_USBD_ITF_TOTAL
};

static const uint8_t const desc_fs_cfg[] = 
{
    // Config number, interface count, string index, total length, attribute, power in mA
    TUD_CONFIG_DESCRIPTOR(1, PICO_USBD_ITF_TOTAL, 0, CONFIG_TOTAL_LEN, 
                          TUSB_DESC_CONFIG_ATT_REMOTE_WAKEUP, PICO_USBD_MAX_POWER_MA),

    // 1st CDC: Interface number, string index, EP notification address and size, EP data address (out, in) and size.
    TUD_CDC_DESCRIPTOR(PICO_USBD_ITF_CDC_0, 4, PICO_EPNUM_CDC_0_NOTIF,
                       PICO_USBD_CDC_CMD_MAX_SIZE, PICO_EPNUM_CDC_0_DATA, 0x80 | PICO_EPNUM_CDC_0_DATA,
                       PICO_USBD_CDC_IN_OUT_MAX_SIZE),

    // 2nd CDC: Interface number, string index, EP notification address and size, EP data address (out, in) and size.
    TUD_CDC_DESCRIPTOR(PICO_USBD_ITF_CDC_1, 4, PICO_EPNUM_CDC_1_NOTIF, 
                       PICO_USBD_CDC_CMD_MAX_SIZE, PICO_EPNUM_CDC_1_DATA, 0x80 | PICO_EPNUM_CDC_1_DATA,
                       PICO_USBD_CDC_IN_OUT_MAX_SIZE),

    // 3rd CDC: Interface number, string index, EP notification address and size, EP data address (out, in) and size.
    TUD_CDC_DESCRIPTOR(PICO_USBD_ITF_CDC_2, 4, PICO_EPNUM_CDC_2_NOTIF, 
                       PICO_USBD_CDC_CMD_MAX_SIZE, PICO_EPNUM_CDC_2_DATA, 0x80 | PICO_EPNUM_CDC_2_DATA,
                       PICO_USBD_CDC_IN_OUT_MAX_SIZE),

    // 4th CDC: Interface number, string index, EP notification address and size, EP data address (out, in) and size.
    TUD_CDC_DESCRIPTOR(PICO_USBD_ITF_CDC_3, 4, PICO_EPNUM_CDC_3_NOTIF, 
                       PICO_USBD_CDC_CMD_MAX_SIZE, PICO_EPNUM_CDC_3_DATA, 0x80 | PICO_EPNUM_CDC_3_DATA,
                       PICO_USBD_CDC_IN_OUT_MAX_SIZE),

    // 5th CDC: Interface number, string index, EP notification address and size, EP data address (out, in) and size.
    TUD_CDC_DESCRIPTOR(PICO_USBD_ITF_CDC_4, 4, PICO_EPNUM_CDC_4_NOTIF, 
                       PICO_USBD_CDC_CMD_MAX_SIZE, PICO_EPNUM_CDC_4_DATA, 0x80 | PICO_EPNUM_CDC_4_DATA,
                       PICO_USBD_CDC_IN_OUT_MAX_SIZE)
};

I think maybe the issue is PICO_EPNUM_CDC_4_DATA = 0x0A, It is over???
I don't understand about usb cdc settings. Please help me review it.[](url)

**Comment by hathach** [comment, marked as answer]
number of endpoints is default to 9, RPI has up to 16, just add this line to tusb_config.h . I make a better default later on.

```C
#define CFG_TUD_EP_MAX 16
````

**Reply by nhhiep92** [reply]
Thanks for your support,  it worked now.

**Reply by hathach** [reply]
Chuyện nhỏ
